net processing: Add support for getcfheaders

jnewbery opened the pull request:
Support `getcfheaders` requests when `-peerblockfilters` is set.

Does not advertise compact filter support in version messages.

jimpo committed:
[test] Add test for cfheaders

jimpo committed:
[net processing] Message handling for getcfheaders.

if -peerblockfilters is configured, handle requests for cfheaders.

jnewbery commented:
This is the 3rd PR taken from #18876. Please see that PR for details of the full changes to support BIP 157.

maflcko reviewed (commented):
Just a style question, feel free to ignore.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -2022,6 +2041,61 @@ static bool PrepareBlockFilterRequest(CNode* pfrom, const CChainParams& chain_pa
+ * @param[in]   connman         Pointer to the connection manager
+ */
+static void ProcessGetCFHeaders(CNode* pfrom, CDataStream& vRecv, const CChainParams& chain_params,
+                                CConnman* connman)
```
```suggestion
                                CConnman& connman)
```

Any reason why this is a pointer, but missing the nullptr check?

* Either this should be a proper pointer with nullptr check, or
* it should be a reference

Mixing the two combines the worst of each option.

(Same for pfrom)

jnewbery replied:
This usage is the same as all the other `Process*()` functions in net_processing. Rather than introduce a new style for just this function, I think it makes sense to change it everywhere in net_processing in a refactor PR.

MarcoFalke replied:
Generally new code should be style-conformant. See for example `RelayTransaction`, which properly passes connman. Obviously a non-blocking nit, so feel free to close this conversation.

jkczyz commented on test/functional/p2p_blockfilters.py:
```diff
@@ -100,12 +104,43 @@ def run_test(self):
+        )
+        node0.send_and_ping(request)
+        response = node0.last_message['cfheaders']
+        assert_equal(
+            compute_last_header(response.prev_header, response.hashes),
+            int(main_cfcheckpt, 16)
+        )
```
Is it worth checking how many hashes are returned? Or would you consider it checked indirectly by `compute_last_header`?

jnewbery replied:
Technically, it's covered indirectly, but I think it's a good idea to explicitly test this so failures are easier to identify. I've added the check.

jkczyz reviewed (commented):
Concept ACK

For testing, what's the general approach been for testing bad requests? BIP 157 says to not respond, but we are disconnecting the peer in these cases.

theStack commented:
Concept ACK

jnewbery committed:
[doc] Add comment for m_headers_cache

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2013,6 +2017,21 @@ static bool PrepareBlockFilterRequest(CNode* pfrom, const CChainParams& chain_pa
         }
     }
 
+    uint32_t stop_height = stop_index->nHeight;
+    if (start_height > stop_height) {
+        LogPrint(BCLog::NET, "peer %d sent invalid getcfilters/getcfheaders with " /* Continued */
+                 "start height %d and stop height %d\n",
+                 pfrom->GetId(), start_height, stop_height);
+        pfrom->fDisconnect = true;
+        return false;
+    }
+    if (stop_height - start_height >= max_height_diff) {
+        LogPrint(BCLog::NET, "peer %d requested too many cfilters/cfheaders: %d / %d\n",
+                 pfrom->GetId(), stop_height - start_height + 1, max_height_diff);
+        pfrom->fDisconnect = true;
+        return false;
+    }
+
```
I think this makes more sense if you also look at the full changes in https://github.com/bitcoin/bitcoin/pull/18876, where this function is common logic for getcfheaders and getcfilters, and the max number of headers that can be requested is 2000, and the max number of filters that can be requested in 1000.

I agree that the getcfcheckpt logic passing in `0` and `std::numeric_limits<uint32_t>::max()` is a bit ugly. I could change it to pass in optionals, but that seems more aesthetic than anything.

theStack replied:
Another option would be to simply use default parameters, i.e. if start_height and max_height diff are not passed, the whole range is allowed. That would need to change the order of the parameters though.

jnewbery commented:
Force pushed the branch to:

- rebase on master now that #18960 is merged
- Added a comment to `m_headers_cache` that didn't make it into #18960 (https://github.com/bitcoin/bitcoin/pull/18960#discussion_r428395728)
- Addressed @jkczyz's comment on the functional test (https://github.com/bitcoin/bitcoin/pull/19010#discussion_r428759826)

fjahr reviewed (approved):
ACK 4376797f6c9962ec92b5ce6e61dc622b56d2a6f7

Reviewed the code, played with the tests, and ran them locally.

FWIW, I have added the tests I suggested in the review club, feel free to copy them from there: https://github.com/fjahr/bitcoin/commit/36d6854d2b4ed0afa556ccbe0ec72c4cb141a4c9

jnewbery commented:
> what's the general approach been for testing bad requests? BIP 157 says to not respond, but we are disconnecting the peer in these cases.

I don't have a strong opinion about whether to disconnect or drop bad requests. Generally I prefer simply dropping bad requests to prevent bugs becoming network partition possibilities, but in this case:

- the message type is entirely new.
- the peers sending us these requests are light clients, so the downside of disconnecting them is small.

I'm inclined to keep it as a disconnect since that's the only way we have to communicate to the client that there's a problem with the request, but I'm happy to change it if others disagree.

MarcoFalke commented:
> the peers sending us these requests are light clients, so the downside of disconnecting them is small.

I agree here. Generally the nicest thing you can do to misbehaving light clients is to disconnect them as early as possible. This way the misbehaviour bug will be hopefully caught early in the design and testing process of the light client. If not, at least in production it will be another indicator that something is wrong and users can file bugs. If the connection was kept alive, and without replying it might just time out on the light client and they'd think *we* are the ones misbehaving.

I was about to suggest to put a small comment in the source code to document this, but I wasn't exactly sure where to put it.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -2048,7 +2122,9 @@ static void ProcessGetCFCheckPt(CNode* pfrom, CDataStream& vRecv, const CChainPa
-    if (!PrepareBlockFilterRequest(pfrom, chain_params, filter_type, stop_hash,
+    if (!PrepareBlockFilterRequest(pfrom, chain_params, filter_type, /*start_height=*/0, stop_hash,
+                                   /*max_height_diff=*/std::numeric_limits<uint32_t>::max(),
+
```
in commit db20d456f0 [net processing] Message handling for getcfheaders.

Why add a empty newline in the params?

jnewbery replied:
Because I have fat fingers. Now fixed.

MarcoFalke commented on src/protocol.h:
```diff
@@ -225,6 +225,17 @@ extern const char* GETBLOCKTXN;
+ * Only available with service bit NODE_COMPACT_FILTERS as described by
+ * BIP 157 & 158.
+ */
+extern const char *GETCFHEADERS;
```
style-nit in commit db20d456f0 (according to clang-format):

```suggestion
extern const char* GETCFHEADERS;
```

jnewbery replied:
Whoops. Missed this out in the rebase. Fixed.

maflcko reviewed (approved):
Some style nits that can be ignored

re-ACK 4376797f6c, only change since my previous review https://github.com/bitcoin/bitcoin/pull/18876#issuecomment-627322015 is adding a spurious new line and a new commit with documentation 🐣

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 4376797f6c, only change since my previous review https://github.com/bitcoin/bitcoin/pull/18876#issuecomment-627322015 is adding a spurious new line and a new commit with documentation 🐣
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjc3gv+MB5VIjvGWjT85TlCzHERqu35XeRFo+DTLBUhkp1NGk1Uymar3cF68jOW
0W0Keide4ZEBCc1sz7mnhsXBBL6XhijSgGS7fTdjw9Iab8944dW11IzDmMFcHWwc
cglhfriS8Rpa3FUeq+FPrNI7zEA4UyHrXcIiUUzXkXBRK0yooTrpFzGMlS6HnB8+
MmzIuUaFTEKnIJA3exIAls8225TzmnIuPrSHtpNi+91RVpTnV1ZMAo++jBbcPE/p
riNswV/Yc/zG06xiO2YgleTyfip8FwjydEMaekkQUSL62IucTFQY0rHqeEWvysFr
L69g2vdnVDUgVLxiOrzLIUdhLjYxsjvyOUdXgC36UxpUERi2TuXzmMu0OWFSkChs
soAvIWL0hFt4oV/PTgBpxJPKrFtWgifwVy3l8iyZ07eVsGr0Z+YfNDwY6DWCJsM/
d+RG0asZEtOjdkwZF/PAc9tTWPsP/fLXiHO3z7928dASyT5+pKdaKOCSCeKH1+6p
lg1V3U+3
=0qw1
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `40b1dadd7d8694410c23fb4d6f045b454bee5f2bf365f49994dc7f713b58f816  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010840b1dadd7d8694410c23fb4d6f045b454bee5f2bf365f49994dc7f713b58f816f01077901fed4b6efcc1e97a5bf3605812a708fff01024e2db56fbdb1853816c177be40dbf7608f0206dd54db98551f745ee69cc028c5d770159457df8d7c9af0bbed14a8a9cbc58e908f1045ec7b51df0081b3c3bbcc3cf6a810083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0107bc40f188839395d07d743913fed35a708f020f810c372f9b620b571c93b2e6cb6b0ee9036e92d5acb3593706df3bc018c1e6a08f1045ec7b51cf00852b3970c9661874a0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f01087cf9d61c3813009bc08df715ffb420908f1045ec7b51cf008fe1b16db896438e20083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

theStack commented on src/protocol.h:
```diff
@@ -225,6 +225,17 @@ extern const char* GETBLOCKTXN;
+extern const char *GETCFHEADERS;
+/**
+ * cfheaders is a response to a getcfheaders request containing a vector of
+ * filter headers for each block in the requested range.
```
nit: technically, it's only one filter header plus a vector of filter hashes

jnewbery replied:
Thanks. I've corrected the comment.

jnewbery commented on src/index/blockfilterindex.h:
```diff
@@ -39,6 +39,7 @@ class BlockFilterIndex final : public BaseIndex
     size_t WriteFilterToDisk(FlatFilePos& pos, const BlockFilter& filter);
 
     Mutex m_cs_headers_cache;
+    //* cache of block hash to filter header, to avoid disk access when responding to getcfcheckpt. */
```
Yes, this is a typo. Fixed!

jnewbery commented:
Thanks for review @MarcoFalke and @theStack. I've addressed all of your review comments, and also removed some redundant wording from the comment for `CFCHECKPOINT` in protocol.h

MarcoFalke commented on src/protocol.h:
```diff
@@ -235,8 +248,6 @@ extern const char* GETCFCHECKPT;
  * cfcheckpt is a response to a getcfcheckpt request containing a vector of
  * evenly spaced filter headers for blocks on the requested chain.
- * Only available with service bit NODE_COMPACT_FILTERS as described by
- * BIP 157 & 158.
```
Why are you removing this?

jnewbery replied:
It doesn't appear in the comments for `cfheaders` and `cfilter`, so I'm removing it from here for consistency. It seems obvious that `NODE_COMPACT_FILTERS` is required since it's a response to `getcfcheckpt`, which requires `NODE_COMPACT_FILTERS`.

MarcoFalke replied:
Fair enough

maflcko reviewed (commented):
re-ACK 5308c97cca , only change is doc related 🗂

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 5308c97cca , only change is doc related 🗂
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUi5SgwAwz0FXCPt6ETbRNPe2chK9XeE0E40zYsq0hAJ/rw7G3/Rk2LyC9TG2GoW
OaJw75Eg80fAu7ImWcTHiUyhKSlIZKJR9QPKC02RImsFZ66UotzFZ0mefsgtkoh0
2bvUJXabSlrOkwt8y12oED2YS0Cgf3OG2DJCPtCGOq9kZUyK+4Ef5stvG6YiYF1/
GZ+WwyO1iL8G8nPZue3239QgMb3bvX6JxIB8q5oipmMzILg8ja7o/0Cxp1QYdxNF
8tK5nZn0kF4xlDaHrusuWIwgNcPGUYlHbCeCuZoXnompyHn/meyTp2yndmEAp72J
ILVkp9Nkzs7o7Q1vAZM7YCWOxHdw7ncDTq0tdcpr9BPtorlHvKWIsUJHvUBTd2el
mlJJaiM5WNq+yISCld8ItK+utub4g95zrfZWyzvs4pLZQTID9tQCtN8U/919zTn1
OUjYFoy+2AmssmJFtM+CkjJgcSAf6KHYcGXJyY4hhqLvhMsmv5SrW1EYJdy6AW5g
YbOfPwLk
=5MJF
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `eed11e440452664f3996d83b9c085e9a4c0ab77e69ea219b1fe7181017bc09a1  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108eed11e440452664f3996d83b9c085e9a4c0ab77e69ea219b1fe7181017bc09a1f010305dc3024db915fabb57ec18d976a8c008fff010b0037fe1b80ae123b429c84dc5a9c00908f1045ec9501af008ec9209cc36c7039a0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010b8bce582eef51b7ef73bc541375cefa508f1045ec9501af0083acb84d15b04d69a0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010c801f6cb22ba79eb3d132fa76ea1f81408f1045ec9501af008e0dd61589357a0f90083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010ca1660919b5d7101d711b44508dcef6608f1045ec9501af008edb036cdb7a8565a0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

theStack reviewed (approved):
ACK 5308c97ccaf0955e5840956bc1636108a43e6f46 :rocket:

jkczyz commented:
ACK 5308c97ccaf0955e5840956bc1636108a43e6f46

jonatack commented on test/functional/p2p_blockfilters.py:
```diff
@@ -100,12 +104,45 @@ def run_test(self):
+        )
+        node0.send_and_ping(request)
+        response = node0.last_message['cfheaders']
+        assert_equal(len(response.hashes), 1000)
```
style nit: replace the multiple values of 1000 and of 16 in this file with well-named constants

jnewbery replied:
The 16 is used to interpret a hex string into an int (RPC return values are strings), so we shouldn't have a constant for that. I tried replacing the 1000 with a constant, but I don't think it made things any clearer.

jonatack commented on test/functional/test_framework/mininode.py:
```diff
@@ -330,6 +332,7 @@ def on_close(self):
     def on_addr(self, message): pass
     def on_block(self, message): pass
     def on_blocktxn(self, message): pass
+    def on_cfheaders(self, message): pass
```
nit: sort here and lines 34 and 72

jnewbery replied:
This was fixed in the following PR.

jonatack commented on src/net_processing.cpp:
```diff
@@ -1983,14 +1985,16 @@ void static ProcessOrphanTx(CConnman* connman, CTxMemPool& mempool, std::set<uin
  * @param[in]   filter_type     The filter type the request is for. Must be basic filters.
+ * @param[in]   start_height    The start height for the request
  * @param[in]   stop_hash       The stop_hash for the request
+ * @param[in]   max_height_diff The maximum number of items permitted to request, as specified in BIP 157
```
I agree with @fjahr's comment and suggested change.

jnewbery replied:
I'll add this in #19070 if I need to retouch that branch

jonatack reviewed (commented):
ACK 5308c97ccaf0955e5840956bc1636108a43e6f46 with some optional style nits for follow-ups. @fjahr's tests at https://github.com/fjahr/bitcoin/commit/36d6854d2b4ed0afa556ccbe0ec72c4cb141a4c9 look worthwhile.
